File: dev/executor/journal.ts
import { serviceName } from "./paths";
import { fail, printCommand, runText, splitLines } from "./utils";

function serviceJournalLines(output: string): string[] {
	return splitLines(output).filter((line) =>
		/systemd\[[0-9]+\]: (Reloading|Reloaded|Starting|Started|Stopping|Stopped|Failed)/.test(
			line,
		),
	);
}

export function showExecutorServiceLogs(since?: string): void {
	const args = [
		"/usr/bin/journalctl",
		"--user",
		"-u",
		serviceName,
		"--no-pager",
	];
	if (since) {
		args.push("--since", since);
	} else {
		args.push("-n", "50");
	}

	printCommand(args);
	const result = runText(args);
	const lines = serviceJournalLines(result.stdout);
	if (lines.length > 0) {
		console.log(lines.join("\n"));
	}
}

export function showInstanceLogs(name: string, since?: string): void {
	const args = [
		"/usr/bin/journalctl",
		"--user",
		"-t",
		`executor/${name}`,
		"--output=cat",
		"--no-pager",
	];
	if (since) {
		args.push("--since", since);
	} else {
		args.push("-n", "20");
	}

	printCommand(args);
	const result = runText(args);
	const output = result.stdout.trimEnd();
	if (output) {
		console.log(output);
	}
}

export function showRecentLogs(name: string): void {
	const logs = runText([
		"/usr/bin/journalctl",
		"--user",
		"-t",
		`executor/${name}`,
		"-n",
		"5",
		"--output=cat",
		"--no-pager",
	]);
	const output = logs.stdout.trimEnd();
	if (output) {
		console.log(output);
	} else {
		console.log("    (no logs)");
	}
}

export function reloadExecutorService(): void {
	const activeArgs = ["/usr/bin/systemctl", "--user", "is-active", serviceName];
	const active = runText(activeArgs).stdout.trim();

	if (active === "active") {
		const args = ["/usr/bin/systemctl", "--user", "reload", serviceName];
		printCommand(args);
		const result = runText(args);
		if (!result.success) {
			if (result.stderr.trim()) {
				console.error(result.stderr.trim());
			}
			fail(`failed to reload ${serviceName}`);
		}
		return;
	}

	const resetArgs = [
		"/usr/bin/systemctl",
		"--user",
		"reset-failed",
		serviceName,
	];
	printCommand(resetArgs);
	runText(resetArgs);

	const startArgs = ["/usr/bin/systemctl", "--user", "start", serviceName];
	printCommand(startArgs);
	const start = runText(startArgs);
	if (!start.success) {
		const output = `${start.stdout}${start.stderr}`.trim();
		if (output) {
			console.error(output);
		}
		fail(`failed to start ${serviceName}`);
	}
}

export function verifyExecutorServiceActive(): void {
	const activeArgs = ["/usr/bin/systemctl", "--user", "is-active", serviceName];
	printCommand(activeArgs);
	const active = runText(activeArgs);
	const state = active.stdout.trim() || "unknown";
	console.log(state);
	if (state === "active") {
		return;
	}

	const statusArgs = [
		"/usr/bin/systemctl",
		"--user",
		"--no-pager",
		"--full",
		"status",
		serviceName,
	];
	printCommand(statusArgs);
	const status = runText(statusArgs);
	const output = `${status.stdout}${status.stderr}`.trim();
	if (output) {
		console.log(output);
	}
	fail("executor service is not active after reload");
}

File: dev/executor/src/journal.rs
use crate::cli::{fail, CliError};
use crate::paths::SERVICE_NAME as PATH_SERVICE;
use crate::runner::{print_command, run_text, split_lines, RunOptions};

pub const SERVICE_NAME: &str = PATH_SERVICE;

pub fn show_recent_logs(name: &str) {
    let args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-t".into(),
        format!("executor/{name}"),
        "-n".into(),
        "5".into(),
        "--output=cat".into(),
        "--no-pager".into(),
    ];
    let logs = run_text(&args, RunOptions::default());
    let output = logs.stdout.trim_end();
    if output.is_empty() {
        println!("    (no logs)");
    } else {
        println!("{output}");
    }
}

pub fn show_instance_logs(name: &str, since: Option<&str>) {
    let mut args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-t".into(),
        format!("executor/{name}"),
        "--output=cat".into(),
        "--no-pager".into(),
    ];
    if let Some(since) = since {
        args.push("--since".into());
        args.push(since.into());
    } else {
        args.push("-n".into());
        args.push("20".into());
    }
    print_command(&args);
    let result = run_text(&args, RunOptions::default());
    let output = result.stdout.trim_end();
    if !output.is_empty() {
        println!("{output}");
    }
}

pub fn reload_executor_service() -> Result<(), CliError> {
    let active_args = vec![
        "/usr/bin/systemctl".into(),
        "--user".into(),
        "is-active".into(),
        SERVICE_NAME.into(),
    ];
    let active = run_text(&active_args, RunOptions::default()).stdout.trim().to_string();
    if active == "active" {
        let args = vec![
            "/usr/bin/systemctl".into(),
            "--user".into(),
            "reload".into(),
            SERVICE_NAME.into(),
        ];
        print_command(&args);
        let result = run_text(&args, RunOptions::default());
        if !result.success {
            let stderr = result.stderr.trim();
            if !stderr.is_empty() {
                eprintln!("{stderr}");
            }
            return Err(fail(format!("failed to reload {SERVICE_NAME}")));
        }
        return Ok(());
    }
    let reset_args = vec![
        "/usr/bin/systemctl".into(),
        "--user".into(),
        "reset-failed".into(),
        SERVICE_NAME.into(),
    ];
    print_command(&reset_args);
    let _ = run_text(&reset_args, RunOptions::default());
    let start_args = vec![
        "/usr/bin/systemctl".into(),
        "--user".into(),
        "start".into(),
        SERVICE_NAME.into(),
    ];
    print_command(&start_args);
    let start = run_text(&start_args, RunOptions::default());
    if !start.success {
        let output = format!("{}{}", start.stdout, start.stderr).trim().to_string();
        if !output.is_empty() {
            eprintln!("{output}");
        }
        return Err(fail(format!("failed to start {SERVICE_NAME}")));
    }
    Ok(())
}

pub fn verify_executor_service_active() -> Result<(), CliError> {
    let active_args = vec![
        "/usr/bin/systemctl".into(),
        "--user".into(),
        "is-active".into(),
        SERVICE_NAME.into(),
    ];
    print_command(&active_args);
    let active = run_text(&active_args, RunOptions::default());
    let state = {
        let trimmed = active.stdout.trim();
        if trimmed.is_empty() {
            "unknown"
        } else {
            trimmed
        }
        .to_string()
    };
    println!("{state}");
    if state == "active" {
        return Ok(());
    }
    let status_args = vec![
        "/usr/bin/systemctl".into(),
        "--user".into(),
        "--no-pager".into(),
        "--full".into(),
        "status".into(),
        SERVICE_NAME.into(),
    ];
    print_command(&status_args);
    let status = run_text(&status_args, RunOptions::default());
    let output = format!("{}{}", status.stdout, status.stderr).trim().to_string();
    if !output.is_empty() {
        println!("{output}");
    }
    Err(fail("executor service is not active after reload"))
}

pub fn show_executor_service_logs(since: Option<&str>) {
    let mut args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-u".into(),
        SERVICE_NAME.into(),
        "--no-pager".into(),
    ];
    if let Some(since) = since {
        args.push("--since".into());
        args.push(since.into());
    } else {
        args.push("-n".into());
        args.push("50".into());
    }
    print_command(&args);
    let result = run_text(&args, RunOptions::default());
    let lines: Vec<_> = split_lines(&result.stdout)
        .into_iter()
        .filter(|line| {
            line.contains("systemd[")
                && (line.contains("Reloading")
                    || line.contains("Reloaded")
                    || line.contains("Starting")
                    || line.contains("Started")
                    || line.contains("Stopping")
                    || line.contains("Stopped")
                    || line.contains("Failed"))
        })
        .collect();
    if !lines.is_empty() {
        println!("{}", lines.join("\n"));
    }
}
